refactor(platform-browser): Throw tree-shakable errors.

JeanMeche opened the pull request:
In a effort to have more tree-shakable errors, this commit transforms every `Error` into tree-shakable `RuntimeError`s

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)


## Does this PR introduce a breaking change?

- [x] No

AndrewKushnir commented on packages/platform-browser/src/browser.ts:
```diff
@@ -8,6 +8,7 @@
 
 import {CommonModule, DOCUMENT, XhrFactory, ɵPLATFORM_BROWSER_ID as PLATFORM_BROWSER_ID} from '@angular/common';
 import {APP_ID, ApplicationConfig as ApplicationConfigFromCore, ApplicationModule, ApplicationRef, createPlatformFactory, ErrorHandler, Inject, InjectionToken, ModuleWithProviders, NgModule, NgZone, Optional, PLATFORM_ID, PLATFORM_INITIALIZER, platformCore, PlatformRef, Provider, RendererFactory2, SkipSelf, StaticProvider, Testability, TestabilityRegistry, Type, ɵINJECTOR_SCOPE as INJECTOR_SCOPE, ...
+import {RuntimeError} from '@angular/core/src/errors';
```
We should update the line to avoid a deep imports (with that, the symbol import can be moved to the previous line as well where other symbols from the `@angular/core` are imported):
```suggestion
import {ɵRuntimeError as RuntimeError} from '@angular/core';
```

AndrewKushnir commented on packages/platform-browser/src/browser/testability.ts:
```diff
@@ -8,13 +8,17 @@
       const testability = registry.findTestabilityInTree(elem, findInAncestors);
-      if (testability == null) {
-        throw new Error('Could not find testability for element.');
+      if (testability == null && ngDevMode) {
+        throw new RuntimeError(
+            RuntimeErrorCode.TESTABILITY_NOT_FOUND, 'Could not find testability for element.');
       }
```
We can probably leave an error (so we don't change the logic) and just tree-shake-away the error message:
```suggestion
      if (testability == null) {
        throw new RuntimeError(
            RuntimeErrorCode.TESTABILITY_NOT_FOUND, NG_DEV_MODE && 'Could not find testability for element.');
      }
```

The `NG_DEV_MODE` can be defined in this file as:

```
const NG_DEV_MODE = typeof ngDevMode === 'undefined' || !!ngDevMode;
```

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -206,7 +209,9 @@ class DefaultDomRenderer2 implements Renderer2 {
-      throw new Error(`The selector "${selectorOrNode}" did not match any elements`);
+      throw new RuntimeError(
+          RuntimeErrorCode.ROOT_SELECTOR_NO_MATCH,
+          ngDevMode && `The selector "${selectorOrNode}" did not match any elements`);
```
Looks great and let's also use `NG_DEV_MODE` as mentioned above.
```suggestion
          NG_DEV_MODE && `The selector "${selectorOrNode}" did not match any elements`);
```

AndrewKushnir commented on packages/platform-browser/src/security/dom_sanitization_service.ts:
```diff
@@ -173,7 +176,9 @@ export class DomSanitizerImpl extends DomSanitizer {
-        throw new Error('unsafe value used in a script context');
+        throw new RuntimeError(
+            RuntimeErrorCode.SANITIZATION_UNSAFE_SCRIPT,
+            ngDevMode && 'unsafe value used in a script context');
```
We should probably use `NG_DEV_MODE` here as well (and declare a const above in this file):
```suggestion
            NG_DEV_MODE && 'unsafe value used in a script context');
```

JeanMeche replied:
I've seen both used in the code base, what is the purpose of that double check with `NG_DEV_MODE` ?

AndrewKushnir replied:
This is a bit inconsistent in the code at this moment (we plan to improve that in the future). There are places of code where `ngDevMode` is guaranteed to be defined (most of the runtime code in core), so we use it without extra checks. In other cases, we check whether the `ngDevMode` is defined first. The unification that we are thinking of (not yet finalized/decided) is to switch to importing the `ngDevMode` from a file (vs having is as a global), but it would have effect on generated code as well, so this is a larger design exploration that we'd need to do in the future.

AndrewKushnir commented on packages/platform-browser/src/errors.ts:
```diff
@@ -0,0 +1,23 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+/**
+ * The list of error codes used in runtime code of the `router` package.
+ * Reserved error code range: 4000-4999.
```
We should update this text (it refers to the Router). I think we can reserve a range between 5000-5500.

JeanMeche committed:
refactor(platform-browser): Throw tree-shakable errors.

In a effort to have more tree-shakable errors, this commit transforms every `Error` into a tree-shakable `RuntimeError`

JeanMeche commented:
I'm a bit surprised to see a 700 bytes increase in size. Any idea why ?

AndrewKushnir commented:
> I'm a bit surprised to see a 700 bytes increase in size. Any idea why ?

Yeah, that's unexpected. I noticed that the `checkNoSyntheticProp` is marked as retained (thus it made it into the golden file). Try reverting the change in that function and push the change, will see if that resolves the situation.

JeanMeche commented:
> > I'm a bit surprised to see a 700 bytes increase in size. Any idea why ?
> 
> Yeah, that's unexpected. I noticed that the `checkNoSyntheticProp` is marked as retained (thus it made it into the golden file). Try reverting the change in that function and push the change, will see if that resolves the situation.

Removing `NG_DEV_MODE` from within `checkNoSyntheticProp` did the job !

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -299,10 +303,12 @@ class DefaultDomRenderer2 implements Renderer2 {
   if (name.charCodeAt(0) === AT_CHARCODE) {
-    throw new Error(`Unexpected synthetic ${nameKind} ${name} found. Please make sure that:
+    throw new RuntimeError(
+        RuntimeErrorCode.UNEXPECTED_SYNTHETIC,
```
```suggestion
        RuntimeErrorCode.UNEXPECTED_SYNTHETIC_PROPERTY,
```

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -206,7 +208,9 @@ class DefaultDomRenderer2 implements Renderer2 {
     if (!el) {
-      throw new Error(`The selector "${selectorOrNode}" did not match any elements`);
+      throw new RuntimeError(
+          RuntimeErrorCode.ROOT_SELECTOR_NO_MATCH,
```
```suggestion
          RuntimeErrorCode.ROOT_NODE_NOT_FOUND,
```

AndrewKushnir commented on packages/platform-browser/src/errors.ts:
```diff
@@ -0,0 +1,23 @@
+  ROOT_SELECTOR_NO_MATCH = 5004,
+  UNEXPECTED_SYNTHETIC = 5005,
+
+  SANITIZATION_UNSAFE_SCRIPT = 5006,
+  SANITIZATION_UNSAFE_RESOURCE_URL = 5007,
+  SANITIZATION_UNEXPECTED_CTX = 5008,
```
We can probably have a sub-range here:
```suggestion
  // Sanitization-related errors (5100-5200 range)
  SANITIZATION_UNSAFE_SCRIPT = 5100,
  SANITIZATION_UNSAFE_RESOURCE_URL = 5101,
  SANITIZATION_UNEXPECTED_CTX = 5102,
```

AndrewKushnir reviewed (commented):
Looks great 👍 Just left a few minor comments.

josephperrott reviewed (approved):
LGTM

Reviewed-for: fw-security

JeanMeche commented:
Last commit fixed the conflicts.

AndrewKushnir reviewed (commented):
@JeanMeche thanks for rebasing this PR!

The changes look good, but since we add an error list to the `platform-browser` package, we'd need to update the `packages/platform-browser/BUILD.bazel` file as well to introduce a rule similar to [this one in core](https://github.com/angular/angular/blob/main/packages/core/BUILD.bazel#L112-L120). That would also require a new golden file that you'd need to generate (try running the `yarn public-api:update` after adding a new build rule).

Thank you.

AndrewKushnir reviewed (approved):
Looks great, thanks @JeanMeche 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented:
[Presubmit](http://test/OCL:528627071:BASE:528627115:1682989397285:5fad6eb1).

AndrewKushnir commented:
**Caretaker note**: it's a bit unclear why PullApprove check doesn't go "green". The PR was reviewed, approved and it's ready for merge.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: fw-core 
Reviewed-for: fw-security

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

pkozlowski-opensource commented:
> Caretaker note: it's a bit unclear why PullApprove check doesn't go "green". The PR was reviewed, approved and it's ready for merge.

@AndrewKushnir PullApprove was missing public api and security bits - added and merging this one.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 80fe08be96633171567e30eb91dd1d23c60a2952.
